PackageReader.GetSupportedFrameworks() is not comprehensive

## Details about Problem

NuGet product used (NuGet.exe | Visual Studio | MSBuild.exe | dotnet.exe): NuGet.Client 5.10.0

## Repro steps and/or sample project

NuGet package System.Runtime 4.3.1 clearly supports net5.0; it even lists specific dependencies for that framework and it can be installed into projects with that framework without issue. However, **calling `GetSupportedFrameworks()` on that package does not return net5.0 nor does it return any framework which is compatible with net5.0** according to `NuGetFrameworkUtility`.

Code to reproduce:

```
var source = new PackageSource("https://api.nuget.org/v3/index.json");
var repository = new SourceRepository(source, new Repository.ProviderFactory().GetCoreV3());
var downloadResource = repository.GetResource<DownloadResource>();
var downloadResult = await downloadResource.GetDownloadResourceResultAsync(
	new PackageIdentity("System.Runtime", new NuGetVersion("4.3.1")),
	new PackageDownloadContext(new SourceCacheContext()),
	SettingsUtility.GetGlobalPackagesFolder(new NullSettings()),
	NullLogger.Instance,
	CancellationToken.None
);
var supportedFrameworks = downloadResult.PackageReader.GetSupportedFrameworks();

// prints "monoandroid10,monotouch10,net45,net462,portable-net45+win8+wp8+wpa81,win8,wp8,wpa81,xamarinios10,xamarinmac20,xamarintvos10,xamarinwatchos10"
Console.WriteLine(string.Join(",", supportedFrameworks.Select(f => f.GetShortFolderName()).OrderBy(f => f)));
// prints "n/a"
Console.WriteLine(NuGetFrameworkUtility.GetNearest(supportedFrameworks, NuGetFramework.Parse("net5.0"), f => f)?.ToString() ?? "n/a");
```

It's non-intuitive, but `net50` expands to a Target Framework Identifier of `.NETCoreApp`, [as per the sepec](https://github.com/dotnet/designs/blob/main/accepted/2020/net5/net5.md), just like `netcoreapp3.1`, `netcoreapp2.1`, and so on do.  On the other hand, `net472` expands to a Target Framework Identifier of `.NETFramework`.  Since `.NETFramework` does not equal `.NETCoreApp`, these two target frameworks are in fact **not** compatible.

The reason that a project using `net5.0` can restore .NET Framework packages, is that the [.NET SDK adds Asset Target Fallback](https://github.com/dotnet/sdk/blob/c53c8ac81e1bcb8118c31bc4f5622396c72e9959/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets#L177-L179), to tell NuGet to ignore compatibility checks and restore anyway. This results in [warning NU1701](https://docs.microsoft.com/nuget/reference/errors-and-warnings/nu1701). As you can see from the SDK targets, projects can disable this implicit fallback by setting a property, so it's not true that any .NETCoreApp project (whether it's netcoreapp3.1, or net5.0, or net6.0) is compatible with .NET Framework packages.

If you're using the NuGet SDK, as you are in your reproduction sample, then you're responsible for mimicking this behaviour.  If you're dealing with real project files, then you can use the MSBuild APIs to read the AssetTargetFallback property, although in a nuget or msbuild restore, targets might modify the value before NuGet is run, making it impossible to have 100% compatibility with customers who customise their build with their own targets.  If you're happy with hardcoding the same fallback TFMs that the SDK does, you can.  Once you have the fallback TFMs you want to use, you can iterate the list after trying the preferred/"project" TFM. Also, you can investigate [`AssetTargetFallbackFramework`](https://github.com/NuGet/NuGet.Client/blob/b102b7587f0286d83dad8ee357caded56bd0ab17/src/NuGet.Core/NuGet.Frameworks/AssetTargetFallbackFramework.cs). Maybe it handles it for you.

@zivkan I'm aware that a .NET framework package is not directly compatible with net5.0. I'm still confused about this specific case, though.

From reading your comment, it sounds like you're saying that System.Runtime 4.3.1 _does not_, in fact, support net5.0. However, you say that this package can be installed into a net5.0 project and will result in a warning NU1701.

However, looking at the package on NuGet, it has an explicit dependency list for netstandard1.0-1.5, all of which _are_ compatible with net 5.0.

To test, I tried making a net5.0 project and installing System.Runtime 4.3.1. *It builds without warnings*. Here's the project file I'm using: 

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <WarningsAsErrors>true</WarningsAsErrors>
    <RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
  </PropertyGroup>

 <ItemGroup>
   <PackageReference Include="System.Runtime" Version="4.3.1" />
 </ItemGroup>
</Project>
```

What am I missing?

--- a/src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs
+++ b/src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs
@@ -360,6 +360,10 @@ namespace NuGet.Packaging
         /// <summary>
         /// Frameworks mentioned in the package.
         /// </summary>
+        /// <remarks>
+        /// This method returns the target frameworks supported for packages.config projects.
+        /// For PackageReference compatibility, use <see cref="NuGet.Client.ManagedCodeConventions"/>
+        /// </remarks>
         public virtual IEnumerable<NuGetFramework> GetSupportedFrameworks()
         {
             var frameworks = new HashSet<NuGetFramework>(new NuGetFrameworkFullComparer());
@@ -377,6 +381,13 @@ namespace NuGet.Packaging
             return frameworks.Where(f => !f.IsUnsupported).OrderBy(f => f, new NuGetFrameworkSorter());
         }
 
+        /// <summary>
+        /// Frameworks mentioned in the package.
+        /// </summary>
+        /// <remarks>
+        /// This method returns the target frameworks supported for packages.config projects.
+        /// For PackageReference compatibility, use <see cref="NuGet.Client.ManagedCodeConventions"/>
+        /// </remarks>
         public virtual Task<IEnumerable<NuGetFramework>> GetSupportedFrameworksAsync(CancellationToken cancellationToken)
         {
             return Task.FromResult(GetSupportedFrameworks());

--- a/src/NuGet.Core/NuGet.Protocol/Plugins/PluginPackageReader.cs
+++ b/src/NuGet.Core/NuGet.Protocol/Plugins/PluginPackageReader.cs
@@ -507,15 +507,7 @@ namespace NuGet.Protocol.Plugins
             throw new NotSupportedException();
         }
 
-        /// <summary>
-        /// Asynchronously gets supported frameworks.
-        /// </summary>
-        /// <param name="cancellationToken">A cancellation token.</param>
-        /// <returns>A task that represents the asynchronous operation.
-        /// The task result (<see cref="Task{TResult}.Result" />) returns an
-        /// <see cref="IEnumerable{NuGetFramework}" />.</returns>
-        /// <exception cref="OperationCanceledException">Thrown if <paramref name="cancellationToken" />
-        /// is cancelled.</exception>
+        /// <inheritdoc cref="PackageReaderBase.GetSupportedFrameworksAsync(CancellationToken)"/>
         public override async Task<IEnumerable<NuGetFramework>> GetSupportedFrameworksAsync(
             CancellationToken cancellationToken)
         {

